### PR TITLE
  Calculated results do not get updated when importing instrument results

### PR DESCRIPTION
**Pull request by Lunga001**
## Description of the issue/feature this PR addresses

Linked issue: https://github.com/senaite/senaite.core/issues/530

## Current behavior before PR

Calculated results do not get updated when importing instrument results

## Desired behavior after PR is merged

Calculated results should updated aswell

--
I confirm I have tested this PR thoroughly and coded it according to [PEP8][1]
and [Plone's Python styleguide][2] standards.

[1]: https://www.python.org/dev/peps/pep-0008
[2]: https://docs.plone.org/develop/styleguide/python.html

### CODE REVIEW

File: bika/lims/exportimport/instruments/resultsimport.py [hunk 1]
```diff
@@ -753,6 +753,44 @@ def _getZODBAnalysesFromReferenceAnalyses(self, objid, criteria):
 
         return analyses
 
+    def analysisHasCalcution(self, objid, analysis):
+        """ returns True or False"""
```
Comment (ramonski) [inline bika/lims/exportimport/instruments/resultsimport.py]: Please write a more descriptive docstring, e.g.:

```python
def analysisHasCalcution(self, objid, analysis):
    """Checks if an Analysis has an calculation

    :param objid: AR ID or Worksheet's Reference Sample IDs
    :param analysis: Analysis Object
    :returns: True if one of the analyses has a calculation
    """
```

Btw, why do we need `analysis` in this method? It is not used at all.

File: bika/lims/exportimport/instruments/resultsimport.py [hunk 2]
```diff
@@ -753,6 +753,44 @@ def _getZODBAnalysesFromReferenceAnalyses(self, objid, criteria):
...
+
+        analyses = self._getZODBAnalyses(objid)
+        analyses_with_calculation = filter(lambda c: c.getCalculation(),analyses)
+        if analyses_with_calculation:
+            return True
```
Comment (ramonski) [inline bika/lims/exportimport/instruments/resultsimport.py]: I think a `for` loop would be better in this case, because we can return True if just one analysis has a calculation attached, e.g.:

```python
for analysis in analyses:
    if analysis.getCalculation():
        return True
return False
```

File: bika/lims/exportimport/instruments/resultsimport.py [hunk 3]
```diff
@@ -753,6 +753,44 @@ def _getZODBAnalysesFromReferenceAnalyses(self, objid, criteria):
...
+        return False
+
+    def getAnalysesWithCalculation(self, objid, analysis, analyses):
+        if self.analysisHasCalcution(objid, analysis):
+            analyses_with_calculation = filter(lambda c: c.getCalculation(),analyses)
+            return analyses_with_calculation
```
Comment (ramonski) [inline bika/lims/exportimport/instruments/resultsimport.py]: This method does the same check code as `analysisHasCalcution` with the difference that it returns the list of analyses instead of casting the list to a boolean. IMO it is not needed at all, because you can do it in `calcluateTotalResults` just at the beginning.

File: bika/lims/exportimport/instruments/resultsimport.py (hunk 2 above)
Comment (ramonski) [reply bika/lims/exportimport/instruments/resultsimport.py]: This method is IMO not needed and can be removed

File: bika/lims/exportimport/instruments/resultsimport.py [hunk 4]
```diff
@@ -753,6 +753,44 @@ def _getZODBAnalysesFromReferenceAnalyses(self, objid, criteria):
...
+
+    def calculateTotalResults(self, objid, analysis):
+        if self.analysisHasCalcution(objid, analysis):
```
Comment (ramonski) [inline bika/lims/exportimport/instruments/resultsimport.py]: Just do that here:
```python
analyses = self._getZODBAnalyses(objid)
analyses_with_calculations = filter(None, map(lambda an: an.getCalculation, analyses))
for an in analyses_with_calculations:
    ...
```

File: bika/lims/exportimport/instruments/resultsimport.py [hunk 5]
```diff
@@ -753,6 +753,44 @@ def _getZODBAnalysesFromReferenceAnalyses(self, objid, criteria):
...
+        if analyses_with_calculation:
+            return True
+        return False
+
+    def getAnalysesWithCalculation(self, objid, analysis, analyses):
+        if self.analysisHasCalcution(objid, analysis):
+            analyses_with_calculation = filter(lambda c: c.getCalculation(),analyses)
+            return analyses_with_calculation
+
+    def calculateTotalResults(self, objid, analysis):
```
Comment (ramonski) [inline bika/lims/exportimport/instruments/resultsimport.py]: Please add a descriptive docstring here (see comment above)

File: bika/lims/exportimport/instruments/resultsimport.py [hunk 6]
```diff
@@ -753,6 +753,44 @@ def _getZODBAnalysesFromReferenceAnalyses(self, objid, criteria):
...
+        if self.analysisHasCalcution(objid, analysis):
+            # Analysis With Calculation
+            analyses = self._getZODBAnalyses(objid)
+            analyses_with_calculation = self.getAnalysesWithCalculation(objid, analysis, analyses)
+            for analysis_with_calc in analyses_with_calculation:
+                calcultion = analysis_with_calc.getCalculation()
+                formular = calcultion.getMinifiedFormula()
+                # The analysis that we are currenly on
+                analysis_keyword = analysis.getKeyword()
+                if analysis_keyword in formular:
```
Comment (ramonski) [inline bika/lims/exportimport/instruments/resultsimport.py]: reduce complexity (level of indendation) and just continue here, e.g.:

```python
formula = calcultion.getMinifiedFormula()
if analysis_keyword not in formula:
    continue
...
```

File: bika/lims/exportimport/instruments/resultsimport.py [hunk 7]
```diff
@@ -753,6 +753,44 @@ def _getZODBAnalysesFromReferenceAnalyses(self, objid, criteria):
...
+                    calc_passed = analysis_with_calc.calculateResult(override=self._override[1])
```
Comment (ramonski) [inline bika/lims/exportimport/instruments/resultsimport.py]: Please add a one line comment above. This helps the reviewer to understand what this piece of code does w/o the need to inspect the whole file. You know the code better, help us to understand what you found out Lunga.

File: bika/lims/exportimport/instruments/resultsimport.py [hunk 8]
```diff
@@ -753,6 +753,44 @@ def _getZODBAnalysesFromReferenceAnalyses(self, objid, criteria):
...
+                    if calc_passed:
+                        #TODO: use api
+                        doActionFor(analysis_with_calc, 'submit')
```
Comment (ramonski) [inline bika/lims/exportimport/instruments/resultsimport.py]: just do it;)

```
from bika.lims import api
api.do_transition_for(analysis_with_calc, "submit")
```

File: bika/lims/tests/doctests/AbbottM2000rtImportInterfaceTotalResults.rst [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+
+And finally check if indeed the analysis has the imported results::
+
+    >>> analyses = ar.getAnalyses()
+    >>> an = [analysis.getObject() for analysis in analyses if analysis.Title == 'HIV06ml'][0]
+    >>> an.getResult()
+    '18'
+    >>> an = [analysis.getObject() for analysis in analyses if analysis.Title == 'Test Total Results'][0]
+    >>> an.getResult()
+    '1800.0'
```
Comment (ramonski) [inline bika/lims/tests/doctests/AbbottM2000rtImportInterfaceTotalResults.rst]: Nice test, well done!

File: bika/lims/exportimport/instruments/resultsimport.py [hunk 10]
```diff
@@ -834,6 +872,8 @@ def _process_analysis(self, objid, analysis, values):
 
         if resultsaved or len(interimsout) > 0:
             doActionFor(analysis, 'submit')
+            self.calculateTotalResults(objid, analysis)
```
Comment (xispa) [inline bika/lims/exportimport/instruments/resultsimport.py]: You are missing to trigger the calculation for the current analysis. `self.calculateTotalResults(objid, analysis)` only performs the calculation of results for the analyses that depends on the current analysis, but not for the analysis itself. One option would be to restore the L810. Another option would be to add `analysis.calculateResult(override=self._override[1])` just after `self.calculateTotalResults(objid, analysis)`

TL;DR
There are Calculations that do not require of results from other Analyses and only use Interim fields. Results for analyses from this type will not be calculated. In fact, the calculation of results for analyses with only interims was correctly performed before your change, but calculation of results for analyses that depend on other analyses were not calculated. Now, happens the other way round.

File: bika/lims/exportimport/instruments/resultsimport.py [hunk 11]
```diff
@@ -824,36 +829,22 @@ def _process_analysis(self, objid, analysis, values):
                 ninterim['value'] = res
                 interimsout.append(ninterim)
                 resultsaved = True
```
Comment (rockfruit) [inline bika/lims/exportimport/instruments/resultsimport.py]: only interims found so far, so resultsaved is true, but.....
